Errors with numpy dev: TypeError: Cannot cast array from dtype

Full log is here: https://travis-ci.org/astropy/astropy/jobs/545541894#L1133, but all failures look the same to the example copied below. The first failing built was 14hours ago.

```
============================= test session starts ==============================
platform linux -- Python 3.7.3, pytest-3.10.1, py-1.8.0, pluggy-0.12.0
Running tests with Astropy version 4.0.dev788.
Running tests in astropy docs.
Date: 2019-06-14T02:30:53
Platform: Linux-4.4.0-101-generic-x86_64-with-debian-jessie-sid
Executable: /home/travis/miniconda/envs/test/bin/python
Full Python Version: 
3.7.3 (default, Mar 27 2019, 22:11:17) 
[GCC 7.3.0]
encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15
Numpy: 1.17.0.dev0+a84476d
Scipy: 1.3.0
Matplotlib: 3.1.0+1179.ge1f0101ef
h5py: 2.9.0
Pandas: 0.24.2
Cython: 0.29.10
Scikit-image: 0.15.0
astropy_helpers: 3.2.1
Using Astropy options: remote_data: any.
rootdir: /tmp/astropy-test-4_t357cj/lib/python3.7/site-packages, inifile: setup.cfg
plugins: remotedata-0.3.1, openfiles-0.3.2, doctestplus-0.3.0, arraydiff-0.3
collected 12840 items / 6 errors                                               
==================================== ERRORS ====================================
_____________ ERROR collecting astropy/io/ascii/tests/test_ecsv.py _____________
astropy/coordinates/earth.py:189: in __new__
    self = cls.from_geocentric(*args, **kwargs)
astropy/coordinates/earth.py:247: in from_geocentric
    return super().__new__(cls, struc, unit, copy=False)
astropy/units/quantity.py:381: in __new__
    value = value.astype(float)
E   TypeError: Cannot cast array from dtype([('x', '<f8'), ('y', '<f8'), ('z', '<f8')]) to dtype('float64') according to the rule 'unsafe'
During handling of the above exception, another exception occurred:
astropy/coordinates/earth.py:192: in __new__
    self = cls.from_geodetic(*args, **kwargs)
E   TypeError: from_geodetic() got an unexpected keyword argument 'x'
During handling of the above exception, another exception occurred:
astropy/io/ascii/tests/test_ecsv.py:257: in <module>
    el = EarthLocation(x=[1, 2] * u.km, y=[3, 4] * u.km, z=[5, 6] * u.km)
astropy/coordinates/earth.py:197: in __new__
    .format(exc_geocentric, exc_geodetic))
E   TypeError: Coordinates could not be parsed as either geocentric or geodetic, with respective exceptions "Cannot cast array from dtype([('x', '<f8'), ('y', '<f8'), ('z', '<f8')]) to dtype('float64') according to the rule 'unsafe'" and "from_geodetic() got an unexpected keyword argument 'x'"
_____________ ERROR collecting astropy/io/ascii/tests/test_ecsv.py _____________
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -319,6 +319,9 @@ astropy.uncertainty
 astropy.units
 ^^^^^^^^^^^^^
 
+- Fixed a bug that prevented ``EarthLocation`` from being initialized with
+  numpy >=1.17. [#8849]
+
 astropy.utils
 ^^^^^^^^^^^^^
 

--- a/astropy/coordinates/angle_utilities.py
+++ b/astropy/coordinates/angle_utilities.py
@@ -787,14 +787,16 @@ def offset_by(lon, lat, posang, distance):
 
     A = Angle(np.arctan2(xsin_A, xcos_A), u.radian)
     # Treat the poles as if they are infinitesimally far from pole but at given lon
-    # The +0*xsin_A is to broadcast a scalar to vector as necessary
-    w_pole = np.argwhere((sin_c + 0*xsin_A) < 1e-12)
-    if len(w_pole) > 0:
+    small_sin_c = sin_c < 1e-12
+    if small_sin_c.any():
         # For south pole (cos_c = -1), A = posang; for North pole, A=180 deg - posang
         A_pole = (90*u.deg + cos_c*(90*u.deg-Angle(posang, u.radian))).to(u.rad)
-        try:
-            A[w_pole] = A_pole[w_pole]
-        except TypeError as e: # scalar
+        if A.shape:
+            # broadcast to ensure the shape is like that of A, which is also
+            # affected by the (possible) shapes of lat, posang, and distance.
+            small_sin_c = np.broadcast_to(small_sin_c, A.shape)
+            A[small_sin_c] = A_pole[small_sin_c]
+        else:
             A = A_pole
 
     outlon = (Angle(lon, u.radian) + A).wrap_at(360.0*u.deg).to(u.deg)

--- a/astropy/units/quantity.py
+++ b/astropy/units/quantity.py
@@ -296,7 +296,8 @@ class Quantity(np.ndarray, metaclass=InheritDocstrings):
                 if not copy:
                     return value
 
-                if not np.can_cast(np.float32, value.dtype):
+                if not (np.can_cast(np.float32, value.dtype) or
+                        value.dtype.fields):
                     dtype = float
 
             return np.array(value, dtype=dtype, copy=copy, order=order,
@@ -376,7 +377,8 @@ class Quantity(np.ndarray, metaclass=InheritDocstrings):
                             "Numpy numeric type.")
 
         # by default, cast any integer, boolean, etc., to float
-        if dtype is None and (not np.can_cast(np.float32, value.dtype)
+        if dtype is None and (not (np.can_cast(np.float32, value.dtype)
+                                   or value.dtype.fields)
                               or value.dtype.kind == 'O'):
             value = value.astype(float)
 

